src/mono/mono/metadata/gc.c

```diff
@@ -879,6 +882,13 @@ finalizer_thread (gpointer unused)
 	mono_hazard_pointer_install_free_queue_size_callback (hazard_free_queue_is_too_big);
 
 	while (!finished) {
+
+		/* Just in case we've received a SIGTERM */
+		if (term_signaled) {
+			mono_runtime_try_shutdown();
+			exit(mono_environment_exitcode_get());
+		}
+
```

Comment:
Not sure if it matters since it's a loop, but maybe move this check after the `finalizer_sem` wait:

```c
  if (wait) {
    mono_coop_sem_wait (&finalizer_sem, ...);
  }
  wait = TRUE;

  if (term_signaled) {
    ...
  }

  mono_thread_info_set_flags (...);
  ...
```

Reply:
Given its the 1st thing to be checked after the thread is given a nudge it probably makes more sense. Though the other finalizer work would be completed before we checked for shutdown. Are there any issues if `mono_run_background_work()` doesn't get done?